lib/compat/wordpress-6.3/template-parts-preloading.php

```diff
@@ -19,17 +19,26 @@ function gutenberg_preload_template_parts( $preload_paths, $context ) {
 		return $preload_paths;
 	}
 
-	$theme_slug = wp_get_theme()->get_stylesheet();
+	// Get any template parts defined in theme.json.
+	$theme_json_template_parts = WP_Theme_JSON_Resolver::get_merged_data()->get_template_parts();
```

Comment:
@tyxla What do you think about this more selective approach? Here we only preload those template parts that are defined **by the Theme** in the Theme JSON.

Reply:
Example

https://github.com/WordPress/twentytwentythree/blob/11440a5b38bae6882fc2c00b703d4bb95420a0f5/theme.json#L712-L733

Reply:
@scruffian I don't really understand the value of having a `templateParts` definition in Theme JSON but Themes are doing it. It could act as a optimising with the Theme "telling" the editor how to optimise itself.

Reply:
This looks like a sensible approach 👍 

I assume that `WP_Theme_JSON_Resolver::get_merged_data()` caches data in runtime so it doesn't have to be processed on every call? 

Have you been able to measure its performance impact? It seems to be that it would load pretty quickly because the theme.json data has already been parsed and is available in the memory, but I wanted to double-check.

Reply:
I haven't manually tried this change, though running the performance tests locally this PR performs worse in every metric but typing :thinking: 

- this PR:

![230320-branch-preloading-template-parts](https://user-images.githubusercontent.com/583546/226353578-bad6b79a-cd44-40cc-81c8-4b6b9ec96677.png)


- trunk

![image](https://user-images.githubusercontent.com/583546/226353328-ad2cd628-0560-4075-9159-a45656937984.png)

Reply:
I now have tested this locally by throttling to "slow 3g" and loading http://localhost:8889/wp-admin/site-editor.php (which means loading "browse mode", not the site editor in edit mode).

In that setup, the canvas contents take ages to load. Within the canvas, the components that took the most time were navigation and query block. The header template was shown almost immediately, both in trunk and this PR.

Perhaps the stats are a bit misleading for this change, because none measure "everything (or the header) has been rendered". First Contentful Paint is probably rendering the site title "Mindblown: a blog about philosophy". I don't see that we track Largest Contentful Paint, though given the importance of the header for TT3 in the whole template, I'd expect it to be affected by this change.

TLDR: I don't see any positive effect of this change. However! If other people are, I don't want to block this. I'll leave space for others to comment.

Reply:
Sounds like you're seeing different things to me. I'll refine the testing criteria and advise further here.